hphp/runtime/vm/jit/vasm-simplify.cpp

```diff
@@ -473,6 +474,48 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
   });
 }
 
+/*
+ * Optimize load followed by truncation
+ * loadq ; movtqb -> loadb (we could use loadtqb as well)
+ * loadq ; movtql -> loadl
+ */
+template<typename Inst>
+bool simplify(Env&, const load& load, const Inst& inst, Vlabel b, size_t i) { return false; }
+
+#define Y(mov_opc, load_opc, be_offset)                              \
+bool simplify(Env& env, const load& load, const mov_opc& mov, Vlabel b, size_t i) { \
+  if (load.d != mov.s) return false;                               \
+  return simplify_impl(env, b, i, [&] (Vout& v) {                  \
+    if (endian() == Endian::LITTLE)                                \
+      v << load_opc{load.s, mov.d};                                \
+    else                                                           \
+      v << load_opc{load.s + be_offset, mov.d};                    \
+    return 2;                                                      \
+  });                                                              \
+}
+
+Y(movtqb, loadb, 7)
+Y(movtql, loadl, 4)
```

Comment:
This pattern is common in `vasm-arm.cpp`, but we actually prefer to avoid it in general.  Rather than macro-define these functions, you should define a templatized implementation, e.g.

```
template<typename Load, typename mov_op>
bool simplify_load_truncate(Env& env, const load& vload, Vlabel b, size_t i) {
  return if_inst<mov_op>(env, b, i + 1, [&] (const op_type<mov_op>& vmov) {
    // check conditions
    return simplify_impl(env, b, i, [&] (Vout& v) {
      ...
      v << Load{...};
      ...
    });
  });
}

bool simplify(Env& env, const load& vload, Vlabel b, size_t i) {
  return
    simplify_load_truncate<loadtqb, Vinstr::movtqb>(env, vload, b, i) ||
    simplify_load_truncate<loadtql, Vinstr::movtql>(env, vload, b, i);
}
```

(It doesn't look like we actually have a `loadtql{}`, but it should be easy to add one.)

I like the caution you're taking around endianness, but if you look at the implementations of `loadtqb{}` among the different architectures, they're all just identical to `loadb{}` currently.  The endianness check should probably be handled there (though I'm sure there are plenty of other things that would break us on big-endian systems today).

For the scope of this PR, I think it would be best to implement `loadtql{}` for all three architectures, copying the implementation of `loadl{}`, and use `loadtqb{}` and `loadtql{}` here in vasm-simplify.  Feel free to put up a separate PR for endianness checks in `vasm-{arm,ppc64,x64}.cpp` if you want; we'd definitely be happy to take those, too.